contributors/design-proposals/pod-priority-api.md

```diff
@@ -0,0 +1,203 @@
+# Priority in Kubernetes API
+
+@bsalamat
+
+May 2017
+  * [Objective](#objective)
+    * [Non-Goals](#oon-Goals)
+  * [Background](#background)
+  * [Overview](#overview)
+  * [Detailed Design](#detailed-design)
+    * [Effect of priority on scheduling](#effect-of-priority-on-scheduling)
+    * [Effect of priority on preemption](#effect-of-priority-on-preemption)
+    * [Priority in PodSpec](#priority-in-podspec)
+    * [Priority names](#priority-names)
+    * [Resolving priority names](#resolving-priority-names)
+    * [Modifying priority names](#modifying-priority-names)
+    * [Ordering of priorities](#ordering-of-priorities)
+    * [Dynamic change of priority names](#dynamic-change-of-priority-names)
+    * [Drawbacks of changing priority names](#drawbacks-of-changing-priority-names)
+    * [Priority and QoS classes](#priority-and-qos-classes)
+
+
+## Objective
+
+
+
+*   How to specify priority for workloads in Kubernetes API.
+*   Define how the order of these priorities are specified. 
+*   Define how new priority levels are added.
+*   Effect of priority on scheduling and preemption.
+
+### Non-Goals 
+
+
+
+*   How preemption works in Kubernetes.
+*   How quota allocation and accounting works for each priority.
+
+## Background 
+
+It is fairly common in clusters to have more tasks than what the cluster
+resources can handle. Often times the workload is a mix of high priority
+critical tasks, and non-urgent tasks that can wait. Cluster management should be
+able to distinguish these workloads in order to decide which ones should acquire
+the resources sooner and which ones can wait. Priority of the workload is one of
+the key metrics that provides the information to the cluster. This document is a
+more detailed design proposal for part of the high-level architecture described
+in [Resource sharing architecture for batch and serving workloads in Kubernetes](https://docs.google.com/document/d/1-H2hnZap7gQivcSU-9j4ZrJ8wE_WwcfOkTeAGjzUyLA).
+
+## Overview 
```

Comment:
This is inconsistent with how we are building similar features in other parts of kubernetes, StorageClasses and ResourceClasses (as currently proposed) come to mind.  Both of these have defined API specs that are validated at creation time and that a UI can reliably build on top of to present the user with the available options. As someone who builds a UI on top of k8s I prefer validated resources to throwing things into ConfigMaps. If this was just a name -> int mapping, maybe a config map.  But its going to be a unique name, possibly a display name, a longer description, and a priority value. So already you have every config map key mapping to a JSON blob.

If you put this kind of information into a ConfigMap you are putting a lot of extra burden on the cluster administrator to make sure their data blobs in the config map are valid.  You will also have to make sure that particular named ConfigMap, in whatever namespace you add it to, has View permissions for every authenticated user in the cluster. Whereas with something like a PriorityClass object it can be properly defined as a cluster scoped resource and you can have the View role generally applied to all PriorityClasses for all users by default.  That also gains you the option in the future to make certain PriorityClasses invisible to certain users.